### DISCUSSION TITLE
how to build complex hashes via build_nested_query and parse them correctly?

### DISCUSSION THREAD

**Discussion by benkoshy** [rack/rack#2158] [Q&A] [answered]
### Use case:

```rb
input = {{hello: "hola", world: "mundo"} => "spanish"}
#=> {{:hello=>"hola", :world=>"mundo"}=>"spanish"}

query = Rack::Utils.build_nested_query(input)
# => "%7B%3Ahello%3D%3E%22hola%22%2C+%3Aworld%3D%3E%22mundo%22%7D=spanish"

# but the output stringifies the key!!
output = Rack::Utils.parse_nested_query(query)
# => {"{:hello=>\"hola\", :world=>\"mundo\"}"=>"spanish"}

# I would like the original hash!
```
### The Goal:

* I have a complex (nested) hash - where some keys are hashes themselves.
* I would like to pass this complex hash - somehow - into a html form and then rebuild / parse it in a controller action. i.e. we need to reliably encode and unencode this.

**Question: Is there any way that this can be done?** Allusion to this [was made here](https://github.com/rack/rack/issues/768#issuecomment-575836958) - but the implementation is not still clear to me.

any pointers would be appreciated.

### Similar discussions

https://github.com/rack/rack/discussions/1974

https://github.com/rack/rack/issues/768#issuecomment-575836958

**Comment by jeremyevans** [comment, marked as answer]
Rack's query parameter serialization format does not support this.  You would have to write your own parameter serializer/deserializer.

**Reply by benkoshy** [reply]
I wrote one, based on a serialization format i made up. Will it be useful to the Rack community or no?

```rb
test "encoding / decoding complex hashes" do 
  	input = {"a"=>{["1","2","3"] =>  [{"name"=>"x"}, {"other"=>"y"}]}}  	
  	query = @coder.build_complex_nested_query(input)
        #=> "0KP=a&0VH-0KA-0KP=0&0VH-0KA-0VP=1&0VH-0KA-1KP=1&0VH-0KA-1VP=2&0VH-0KA-2KP=2&0VH-0KA-2VP=3&0VH-0VA-0KP=0&0VH-0VA-0VH-0KP=name&0VH-0VA-0VH-0VP=x....."
  	assert_equal input, @coder.parse_complex_nested_query(query)
  end
```

**Reply by ioquatix** [reply]
Such features can be useful, but the bar for usefulness often doesn't extend beyond what a browser could produce from `<form>` element unfortunately.

**Reply by benkoshy** [reply]
> Such features can be useful, but the bar for usefulness often doesn't extend beyond what a browser could produce from <form> element unfortunately.

@ioquatix thx for your answer. But I do not understand it. Perhaps I shall ask explicitly: 

* Should I make a PR for this use-case on rack? Would it be useful to anyone (apart from myself)?

My use case was to pass in a complex hash, via a html form and have it come out reliably on the other side without writing custom serialisation / deserialization logic.

My use case was this:

```rb
<%= hidden_field_tag "retail_stock[part_list]", @encoded_complex_hash  %>

# controller_action.rb
@decoded_complex_hash
```

**Reply by ioquatix** [reply]
Sorry if it was not clear.

To answer your question directly, no, I don't think we can accept it. It's too specific.

However, I do believe you can make a gem or library for this feature. I'm not against listing "See Also" gems.

With regards to your specific use case, maybe consider https://github.com/hopsoft/universalid
